### ISSUE TITLE
[Bug] CollectionView infinite scroll does not work with a grouped list on iOS

### ISSUE THREAD

**Issue by kramer-e** [xamarin/Xamarin.Forms#8383]
### Description

When enabling grouping and setting RemainingItemsThreshold on a CollectionView on iOS RemainingItemsThresholdReachedCommand is only executed once. After that RemainingItemsThresholdReached and RemainingItemsThresholdReachedCommand are not called when the threshold is reached. On Android the CollectionView behaves as expected.

### Steps to Reproduce

1. Use the reproduction project, or set RemainingItemsThreshold to 0 and IsGrouped True, load Collectionview with test items and bind RemainingItemsThresholdReachedCommand to a command that loads more test items.
2. Run the iOS project on an iOS simulator.
3. Scroll down and observe that only one group with items is loaded

### Expected Behavior

When scrolling down, new items should automatically load when reaching the end of the list, like on Android.

### Actual Behavior

When scrolling down, only one group is loaded, when scrolling again to the end of the list no subsequent groups are loaded. This occurs on the iOS simulator and on an iPhone 8 with iOS 13.2.

### Basic Information

- Version with issue: 4.5.0.1336-nightly, 4.3.0.947036 (latest stable when submitting issue)
- Last known good version: N/A
- IDE: VS for Mac
- Platform Target Frameworks:
  - iOS:  13.2
  - Android: 9.0 (API level 28)
- Android Support Library Version: 28.0.0.3
- Nuget Packages: Xamarin.Essentials, Xamarin.Forms
- Affected Devices: iPhone 8

### Reproduction Link

https://github.com/kramer-e/CollectionViewInfiniteScrollingBug

**Comment by tranb3r** [comment]
Same for me.
Any update?

**Comment by RSchipper** [comment]
Recently I ran into the same problem as described here. Upon debugging / looking at the source I figured out the event is triggered by the "scrolled" method inside "ItemsViewDelegator" class. This method does not take the grouping in account when calculating the "*itemIndex". Therefor the switch statement responsible for triggering the "itemsThreshold" event will never produce anything usable. My work-around is as follows:

**Custom renderer** 
```csharp
  public class CustomCollectionViewRenderer : CollectionViewRenderer
    {
        protected override GroupableItemsViewController<GroupableItemsView> CreateController(GroupableItemsView itemsView, ItemsViewLayout layout)
        {
            return new CustomGroupableItemsViewController<GroupableItemsView>(itemsView, layout);
        }
    }
```

**Custom controller**
```csharp
    public class CustomGroupableItemsViewController<TItemsView> : GroupableItemsViewController<TItemsView>
        where TItemsView : GroupableItemsView
    {
        public CustomGroupableItemsViewController(TItemsView groupableItemsView, ItemsViewLayout layout)
            : base(groupableItemsView, layout)
        {
        }

        protected override UICollectionViewDelegateFlowLayout CreateDelegator()
        {
            return new CustomGroupableItemsViewDelegator<TItemsView, CustomGroupableItemsViewController<TItemsView>>(ItemsViewLayout, this);
        }
    }
```

**Custom delegator**
```csharp
    public class CustomGroupableItemsViewDelegator<TItemsView, TViewController> : GroupableItemsViewDelegator<TItemsView, TViewController>
        where TItemsView : GroupableItemsView
        where TViewController : GroupableItemsViewController<TItemsView>
    {
        public CustomGroupableItemsViewDelegator(ItemsViewLayout itemsViewLayout, TViewController itemsViewController)
            : base(itemsViewLayout, itemsViewController)
        {
        }

        public override void Scrolled(UIScrollView scrollView)
        {
            List<NSIndexPath> indexPathsForVisibleItems =
                ViewController.CollectionView.IndexPathsForVisibleItems.OrderBy(x => x.Row).ToList();

            if (indexPathsForVisibleItems.Count == 0)
                return;

            TItemsView itemsView = ViewController.ItemsView;
            IItemsViewSource source = ViewController.ItemsSource;

            UIEdgeInsets contentInset = scrollView.ContentInset;
            nfloat contentOffsetX = scrollView.ContentOffset.X + contentInset.Left;
            nfloat contentOffsetY = scrollView.ContentOffset.Y + contentInset.Top;

            UICollectionView collectionView = ViewController.CollectionView;
            CGPoint centerPoint = new CGPoint(collectionView.Center.X + collectionView.ContentOffset.X,
                collectionView.Center.Y + collectionView.ContentOffset.Y);

            NSIndexPath centerIndexPath = collectionView.IndexPathForItemAtPoint(centerPoint);

            NSIndexPath firstVisibleItem = indexPathsForVisibleItems.First();
            NSIndexPath centerItem = centerIndexPath ?? firstVisibleItem;
            NSIndexPath lastVisibleItem = indexPathsForVisibleItems.Last();

            int firstVisibleItemIndex = GetItemIndexWithGrouping(firstVisibleItem, source);
            int centerItemIndex = GetItemIndexWithGrouping(centerItem, source);
            int lastVisibleItemIndex = GetItemIndexWithGrouping(lastVisibleItem, source);

            ItemsViewScrolledEventArgs itemsViewScrolledEventArgs = new ItemsViewScrolledEventArgs
            {
                HorizontalDelta = contentOffsetX - PreviousHorizontalOffset,
                VerticalDelta = contentOffsetY - PreviousVerticalOffset,
                HorizontalOffset = contentOffsetX,
                VerticalOffset = contentOffsetY,
                FirstVisibleItemIndex = firstVisibleItemIndex,
                CenterItemIndex = centerItemIndex,
                LastVisibleItemIndex = lastVisibleItemIndex
            };

            itemsView.SendScrolled(itemsViewScrolledEventArgs);

            PreviousHorizontalOffset = (float)contentOffsetX;
            PreviousVerticalOffset = (float)contentOffsetY;

            switch (itemsView.RemainingItemsThreshold)
            {
                case -1:
                    return;
                case 0:
                    if (lastVisibleItemIndex == source.ItemCount - 1)
                        itemsView.SendRemainingItemsThresholdReached();
                    break;
                default:
                    if (source.ItemCount - 1 - lastVisibleItemIndex <= itemsView.RemainingItemsThreshold)
                        itemsView.SendRemainingItemsThresholdReached();
                    break;
            }
        }

        private static int GetItemIndexWithGrouping(NSIndexPath indexPath, IItemsViewSource itemSource)
        {
            int index = (int)indexPath.Item;

            if (indexPath.Section > 0)
            {
                for (int i = 0; i < indexPath.Section; i++)
                {
                    index += itemSource.ItemCountInGroup(i);
                }
            }

            return index;
        }
    }
```

If it's useful/wanted I can make a PR for this repository, but I'm not sure it's the best fix possible. :)

**Comment by kramer-e** [comment]
Thank you for investigating and sharing the workaround, I just tested it and works. @rmarinho  I hope someone from the Forms team can verify this fix.

**Comment by tranb3r** [comment]
This kind of code should be in a common class shared by the different platforms.

**Comment by jfversluis** [comment]
@RSchipper if you haven't done so already, feel free to open a PR. That will make it easier for us to test this and verify the actual fix since we can just look at the code as it is integrated in our solution instead of having to copy/paste and figure it out ourselves. Thanks :)

**Comment by tranb3r** [comment]
@jfversluis Do you have any idea when this is going to be fixed ? Thanks

**Comment by developer9969** [comment]
hi there i am interested in this one too any news?

**Comment by tranb3r** [comment]
@samhouts Any progress ?

**Comment by tranb3r** [comment]
Any update ?

**Comment by demoh2019** [comment]
Any Update?

**Comment by winterdouglas** [comment]
The fact is that nothing works when the list is grouped, and no one seems to care about it. This issue has been opened in 2019, back then I used the suggested workaround and it kind of helped, but did not solve the issues completely.

**Comment by giuseppenovielli** [comment]
@hartez and other xamarin forms team give a feedback, please Thanks!

**Comment by brunonocetti** [comment]
Any Update?

**Comment by giuseppenovielli** [comment]
?

**Comment by giuseppenovielli** [comment]
@samhouts @hartez @jsuarezruiz @Redth This issue is Hight Impact!
If i want grouped CollectionView and Infinite Scrool, i can't use CollectionView!

Please provide a fix!

Thanks!

**Comment by kramer-e** [comment]
The workaround shared by RSchipper above works, so you can implement that to use infinite scrolling on iOS. I'm sure that the Xamarin team will implement a fix shortly.

**Comment by OscarHMG** [comment]
@RSchipper Could you share a repo with the workaround to fix the Infinite scrolling in Xamarin IOS. In android, I have no issues with this feature but in IOS, doesnt perform the infinite scroll

Thanks in advance.

**Comment by AdrianoBinhara** [comment]
any updates???

**Comment by michaldivis** [comment]
Same problem here, any updates?

**Comment by justintoth** [comment]
@jfversluis @jsuarezruiz I see that you merged the PR into this 5.0.0 SR6 release back in Oct 11. I'm still not seeing this in the latest Xamarin.Forms *Preview* update, 5.0.0.2291. When will this be available in the Preview channel? My CollectionView with 3,000 items is eagerly awaiting the ability to page...

**Comment by jfversluis** [comment]
@justintoth not sure what the preview version has to do with this or where you're getting that from? Since it's part of Xamarin.Forms since SR6, it is now available in the current stable version. If something is still doesn't seem to be working for you, please open a new issue.
